modules/cas_cache/layer_cache_management.c

```diff
@@ -429,58 +440,64 @@ static void cache_start_rollback(struct work_struct *work)
 	if (mngt_queue)
 		ocf_queue_put(mngt_queue);
 
-	module_put(THIS_MODULE);
-
 	result = _cache_mngt_async_callee_set_result(&ctx->async,
 			ctx->ocf_start_error);
 
 	if (result == -KCAS_ERR_WAITING_INTERRUPTED)
 		kfree(ctx);
+
+	module_put_and_exit(0);
+
+	return 0;
 }
 
 static void _cache_mngt_cache_stop_rollback_complete(ocf_cache_t cache,
 		void *priv, int error)
 {
 	struct _cache_mngt_attach_context *ctx = priv;
+	struct task_struct *thread;
 
 	if (error == -OCF_ERR_WRITE_CACHE)
 		printk(KERN_WARNING "Cannot save cache state\n");
 	else
 		BUG_ON(error);
 
-	INIT_WORK(&ctx->work, cache_start_rollback);
-	schedule_work(&ctx->work);
+	thread = kthread_run(cache_start_rollback, ctx,
+			"cas_cache_rollback_complete");
+	BUG_ON(IS_ERR(thread));
 }
```

Comment:
Consider moving thread creation to cache initialization phase to avoid unrecoverable errors here.